src/app/features/signup/signup.page.ts

```diff
@@ -192,73 +192,79 @@ export class SignupPage {
-            err.error.error?.type === 'invalid_referral_code'
-          ) {
-            return this.errorService.toastError$(
-              this.translocoService.translate(
-                'error.diaBackend.invalid_referral_code'
-              )
-            );
+    const device$ = this.diaBackendAuthService.readDevice$();
+    const createUser$ = device$.pipe(
+      concatMap(([fcmToken, deviceInfo, deviceId]) => {
+        return this.diaBackendAuthService.createUser$(
+          this.model.username,
+          this.model.email,
+          this.model.password,
+          this.model.referralCodeOptional,
+          {
+            fcm_token: fcmToken,
+            platform: deviceInfo.platform,
+            device_identifier: deviceId.uuid,
           }
-          // eslint-disable-next-line @typescript-eslint/no-magic-numbers
-          if (err instanceof HttpErrorResponse && err.status === 401)
-            return this.errorService.toastError$(
-              this.translocoService.translate(
-                'error.diaBackend.untrusted_client'
-              )
-            );
-          return this.errorService.toastError$(err);
-        })
-      );
-
+        );
+      })
+    );
+    const action$ = createUser$.pipe(
+      first(),
+      concatMapTo(
+        defer(() =>
+          this.router.navigate(
+            [
+              '/login',
+              { email: this.model.email, password: this.model.password },
+            ],
+            { replaceUrl: true }
+          )
+        )
+      ),
+      catchError((err: unknown) => {
+        return this.handleOnSubmitError(err);
+      })
+    );
     this.blockingActionService
       .run$(action$)
       .pipe(untilDestroyed(this))
       .subscribe();
   }
+
+  private handleOnSubmitError(err: unknown) {
+    if (
+      err instanceof HttpErrorResponse &&
+      err.error.error?.type === 'duplicate_email'
+    ) {
+      return this.errorService.toastError$(
+        this.translocoService.translate('error.diaBackend.duplicate_email')
+      );
+    }
+    if (
+      err instanceof HttpErrorResponse &&
+      err.error.error?.details?.username?.length > 0
+    ) {
+      return this.errorService.toastError$(
+        this.translocoService.translate('error.diaBackend.duplicate_username')
+      );
+    }
+    if (
+      err instanceof HttpErrorResponse &&
+      err.error.error?.type === 'invalid_referral_code'
+    ) {
+      return this.errorService.toastError$(
+        this.translocoService.translate(
+          'error.diaBackend.invalid_referral_code'
+        )
+      );
+    }
+    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
+    if (err instanceof HttpErrorResponse && err.status === 401)
+      return this.errorService.toastError$(
+        this.translocoService.translate('error.diaBackend.untrusted_client')
+      );
+    return this.errorService.toastError$(err);
+  }
 }
 
 interface SignupFormModel {
```

Comment:
The code in this commit sends device information during the sign-up process. You've added a `concatMap` to chain the `device$` observable with the `createUser$` observable, which now includes the `fcm_token`, `platform`, and `device_identifier` in the request. The error handling has been moved to the `handleOnSubmitError` function to improve readability.

Here are a few suggestions for further improvements:

1.  Add comments to explain the purpose of the different observables, e.g., `device$`, `createUser$`, and `action$`.

2.  Change the `if` statements in the `handleOnSubmitError` function to a `switch` statement based on the error type, which can simplify the code and make it more readable.

```typescript

private handleOnSubmitError(err: unknown) {
  if (err instanceof HttpErrorResponse) {
    switch (err.error.error?.type) {
      case 'duplicate_email':
        return this.errorService.toastError$(
          this.translocoService.translate('error.diaBackend.duplicate_email')
        );
      case 'invalid_referral_code':
        return this.errorService.toastError$(
          this.translocoService.translate('error.diaBackend.invalid_referral_code')
        );
      // ...
    }

    if (err.error.error?.details?.username?.length > 0) {
      // ...
    }

    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    if (err.status === 401) {
      // ...
    }
  }

  return this.errorService.toastError$(err);
}
```

Other than these suggestions, the changes made in this commit look good and effectively implement the feature to send device information during the sign-up process.

Reply:
1. SKIP:
2. SKIP: opinionated, if feels better.